ntp-udp/src/raw_socket.rs

```diff
@@ -0,0 +1,386 @@
+
+    use super::{cerr, TimestampingConfig};
+
+    /// Makes the kernel return the timestamp as a cmsg alongside an empty packet,
+    /// as opposed to alongside the original packet
+    const SOF_TIMESTAMPING_OPT_TSONLY: u32 = 1 << 11;
+    /// Makes the kernel return a packet id in the error cmsg.
+    const SOF_TIMESTAMPING_OPT_ID: u32 = 1 << 7;
+
+    pub(crate) fn set_timestamping_options(
+        udp_socket: &std::net::UdpSocket,
+        timestamping: TimestampingConfig,
+    ) -> std::io::Result<()> {
+        let fd = udp_socket.as_raw_fd();
+
+        let mut options = 0;
+
+        if timestamping.rx_software || timestamping.tx_software {
+            // enable software timestamping
+            options |= libc::SOF_TIMESTAMPING_SOFTWARE
+        }
+
+        if timestamping.rx_software {
+            // we want receive timestamps
+            options |= libc::SOF_TIMESTAMPING_RX_SOFTWARE
+        }
+
+        if timestamping.tx_software {
+            // - we want send timestamps
+            // - return just the timestamp, don't send the full message along
+            // - tag the timestamp with an ID
+            options |= libc::SOF_TIMESTAMPING_TX_SOFTWARE
+                | SOF_TIMESTAMPING_OPT_TSONLY
+                | SOF_TIMESTAMPING_OPT_ID;
+        }
+
+        // for documentation on SO_TIMESTAMPING see
+        // https://www.kernel.org/doc/Documentation/networking/timestamping.txt
+        // Safety:
+        // we have a reference to the socket, so fd is a valid file descriptor for the duration of the call
+        // SOL_SOCKET + SO_TIMESTAMPING expect a *u32 as value, which &options is. Furthermore, we own options
+        // hence the pointer is valid for the duration of the call.
+        // option_len is set to the size of u32, which is the size for which the value pointer is valid.
+        unsafe {
+            cerr(libc::setsockopt(
+                fd,
+                libc::SOL_SOCKET,
+                libc::SO_TIMESTAMPING,
+                &options as *const _ as *const libc::c_void,
+                std::mem::size_of::<u32>() as libc::socklen_t,
+            ))?
+        };
+
+        Ok(())
+    }
+}
+
+mod timestamping_config {
+    use std::os::unix::prelude::AsRawFd;
+
+    use super::cerr;
+    use crate::interface_name;
+
+    #[derive(Debug, Clone, Copy, Default)]
+    pub(crate) struct TimestampingConfig {
+        pub(crate) rx_software: bool,
+        pub(crate) tx_software: bool,
+    }
+
+    #[repr(C)]
+    #[allow(non_camel_case_types)]
+    #[derive(Default)]
+    struct ethtool_ts_info {
+        cmd: u32,
+        so_timestamping: u32,
+        phc_index: u32,
+        tx_types: u32,
+        tx_reserved: [u32; 3],
+        rx_filters: u32,
+        rx_reserved: [u32; 3],
+    }
+
+    #[repr(C)]
+    #[allow(non_camel_case_types)]
+    struct ifreq {
+        ifrn_name: [u8; 16],
+        ifru_data: *mut libc::c_void,
+        __empty_space: [u8; 40 - 8],
```

Comment:
just as a note: this ties us to 64-bit operating systems, and also probably to linux. The union that libc uses here is extremely nasty. We should maybe look into upstreaming this?

Reply:
Not sure what you mean here. I did not change anything here so maybe we should fix this separately?

Reply:
I know. It's just that with our eye on supporting more platforms, this is one place where I know we make an assumption about running on a 64-bit OS